lib/elixir/lib/kernel.ex

```diff
@@ -2198,6 +2198,33 @@ defmodule Kernel do
     :erlang.error(ArgumentError.exception(error_message))
   end
 
+  @doc """
+  Returns true if `term` is a struct; otherwise returns `false`.
+
+  Allowed in guard tests.
+  """
+  @doc guard: true
+  defmacro is_struct(term) do
```

Comment:
out of curiosity, why is this implementation preferred over:

```elixir
  defguard is_struct(term)
           when is_map(term) and :erlang.is_map_key(:__struct__, term) and
                  is_atom(:erlang.map_get(:__struct__, term))
```

is it more efficient in some cases?

Or because:

```
# foo.exs
is_struct(42)
```

warns:

```
warning: this expression will fail with BadMapError
  foo.exs:1
```

?

Or both? Or something else entirely?

Reply:
To avoid warning, as it would provide inconsistent behaviour when compared to the rest of `is_*/1` functions.

Reply:
gotcha, thanks!